<a href="https://colab.research.google.com/github/LKXoro/BEDU-Rstudio-postworks/blob/main/Proyecto_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Postwork 01: Identificación del problema

## Estado del arte de los eSports

Los deportes electrónicos o _**eSports**_ como se les conoce hoy en día, se pueden definir como la competencia originada en videojuegos digitales ante una audiencia (Hodge et al., 2017). Con una base de consumidores global cercana a los 450 millones de personas y beneficios de miles de millones de dólares anuales, el crecimiento de los eSports no muestra signos de detenerse y se encuentran un momento muy saludable. De acuerdo con reportes recientes acerca del estado de los eSports, durante Marzo de 2019, los cinco juegos más vistos en Twitch fueron:

*   League of Legends
*   Fortnite
*   DOTA 2
*   Counter Strike: Global Offensive
*   Overwatch

En promedio, estos juegos juntos logran hasta 400 mil expectadores a cualquier hora. Otros reportes indican que los eSports lograrán audiencias mayores, en países como Estados Unidos, que en cualquier liga deportiva a excepción de la NFL (Iqbal, 2019).

<p align="center">
<img src="https://educowebmedia.blob.core.windows.net/educowebmedia/educospain/media/images/blog/que-son-los-esports-portada.jpg"  height ="250" width = "400" alt=eSports>
</p>


# Oportunidades que ofrecen los eSports

Detrás de este crecimiento, encontramos un ecosistema próspero constituido por desarrolladores, jugadores y entrenadores, creadores de contenido, plataformas de consumos y muchos más. A medida que ese ecosistema florece, también lo hacen las oportunidades laborales en el mercado de los eSports, siendo una parte los que se comprometen al análisis estadístico de eSports.

La demanda de analístas de eSports se espera incremente significativamente en los próximos años, de la misma manera que la demanda de analístas para deportes convencionales aumento en 2004 (Davenport, 2014). Previendo este incremento, algunas universidades han decidido preparar a sus estudiantes mediante carreras enfocadas al analísis de eSports. 

<p align="center">
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/cd/National_Association_of_Collegiate_Esports_logo.svg/1200px-National_Association_of_Collegiate_Esports_logo.svg.png" height ="250" width = "400" alt=NACE_logo>
</p>

La Universidad de Akron es por lejos la mayor inversionista en términos de programas e instalaciones enfocadas a eSports, junto a la Asociación Nacional de Esports Universitario ([NACE](https://nacesports.org/)) que, hasta Mayo de 2019, se compone de 144 miembros. A medida que las universidades destinan más recursos a los eSports, la industria global continuara expandiéndose. 

# Información necesaria para el análisis del desempeño en eSports

Muchos grupos de interés se ven en la necesidad de tener información proveniente de análisis de eSports, por ejemplo, los desarrolladores de videojuegos buscan información que les permita encontrar patrones en la jugabilidad con la finaldiad de realizar cambios en su juego como lo son mapas, reglas, personajes u objetos (Bozorgzadeh, 2017). Ellos también buscan detectar trampas y de esta forma crear sistemas "_anti-cheats_" que permitan un juego más justo.

Si bien mucha de la información recolectada para su análisis es explotada para fines como el aumento de audiencia, momentos destacados o para pláticas de comentaristas durante y después de las partidas. Un ámbito que puede ser de mayor intéres es el relacionado con el rendimiento de los equipos y jugadores. Por ejemplo, análistas, coaches e incluso los propios jugadores presentan particular interés en el desempeño de sus adversarios, por lo que al igual que en los deportes tradicionales, deben conocer a su enemigo y preparar sus estrategias.

En general nos encontramos con cuatro aspectos de posible análisis en los eSports:

*   **Equipos**: La mayoría de profesionales de eSports participan en juegos que presentan enfrentamientos entre qeuipos, donde el entendimiento de los oponentes y su estrategia juegan un papel crítico.
*   **Jugadores**: Las debilidades y fortalezas de cada integrante del equipo permiten trazar un plan que permita bloquear su desempeño en el rol que ejecuta.
*   **Personajes**: Los roles de cada jugador en los distintos juegos vienen dados por el avatar o personaje elegido, entender las habilidades de cada personaje y sus limitaciones tienen un peso en el desempeño como jugador.
*   **Objetos**: Cada personaje puede equiparse con distintos objetos, entender las diferentes composicions y como adecuarlas a la situación en que encuentras trae un abánico de posibiliades para el desarrollo de una partida.

Está claro que cada uno de estos aspectos brinda un cúmulo de información por partida que solo puede ser explotada por análistas expertos. Hoy en día también se habla de páginas de apuestas, eSports de fantasía, ligas colegiales y más importante aún: Canales de televisión enteramente dedicados a eSports. Detrás de todo esto surge una pregunta importante: ¿Dónde conseguimos la información? Para responder está pregunta podemos empezar por conocer algunas fuentes de esta información que hoy en día es tan preciada.


# Fuentes de datos sobre eSports

Antes de que la información se analizada, debe ser recolectada. Algunas de las fuentes más comúnes donde la información sobre eSports puede ser recolectada son:

*   **APIs**: Para los juegos más importantes, los desarrolladores proveen una interfaz de programación de aplicaciones ([API](https://es.wikipedia.org/wiki/Interfaz_de_programaci%C3%B3n_de_aplicaciones)) con la cual se permite la descarga de datos generados a los usuarios registrados.
*   **Proveedores estadísticos y APIs de terceros**: Ya que no todos los desarrolladores de juegos recolectan los datos, terceros proveen APIs donde guardan sus datos obtenidos mediante la búsqueda en registros de los juegos, analizando repositorios públicos, haciendo _scrapping_ en repeticiones de Twitch o Youtube, o bien pidiendo archivos de repetición a los usuarios.
*   **Repositorios de archivos**: Como mencionamos anteriormente muchos juegos guarda repeticiones de las partidas, pero no necesariamente en formatos de video, sino en formatos especiales que guardan cada acción de una partida y pueden reproducirse de nuevo con el motor del juego para analizar distintos momentos o ver como evoluciona la partida desde uno o múltiples ángulos. Estos archivos pueden ser tratados o simplemente guardarse en crudo en repositorios.

<p align="center">
<img src="https://blog.ehcgroup.io/wp-content/uploads/2019/08/What-is-an-API.png" height ="250" width = "400" alt=API>
</p>



# Uso de los datos para crear visualizaciones y Dashboards

Otra pregunta que puede surgir es: ¿Qué tipo de trabajo realizan los análistas de eSports? Para responder esta pregunta, se analizaron 22 ofertas para puestos de trabajos como analistas de eSports (Joseph Rubleske et al.). Los análisis revelaron que, mayormente, los trabajos implican dos amplias áreas de trabajo analítico:

*   **Creación de atributos para visualizaciones y dashboards**: Un oferta de **Blizzard** describía sus necesidades como: "_Contar historias con datos y visualizaciones_".
*   **Modelado para Machine Learning**: Construcción de modelos que permitan evaluar las fortalezas y debilidades de los oponentes, predecir el efecto de balanceo de personajes en el ratio de victoria de los equipos, recomendar objetos de acuerdo a la forma de jugar del rival, etc. 

<p align="center">
<img src="https://i.pinimg.com/originals/c9/2d/0c/c92d0c88656b0677c6bc69312769f854.png" height ="250" width = "400" alt=API>

</p>


# El acceso a la información sobre eSports

Ya hemos hablado de las posibles fuentes de datos y que se puede hacer con ellos, pero para los usuarios más entusiastas el acceso a la información en crudo puede resultar complicado. Si bien no todos los desarrolladores de videojuegos están dispuestos a proporcionar información de sus juegos tan fácilmente, no todos los que la comparten a través de una API brindan una accesibilidad y herramientas fáciles de usar, como por ejemplo, lo hace Riot Games con su [API pública](https://developer.riotgames.com/) que brinda acceso a datos de todos sus juegos.

En este punto intermedio entre los datos en crudo y las visualizaciones atractivas nos encontramos con el problema a tratar: El acceso a la información sobre eSports es como ir a pescar en mar abierto. Podemos encontrar muchas APIs que ofrecen información en crudo o muchas páginas web que nos ofrecen sus propios datos recolectados a cambio de un módico precio, otras tantas son más herméticas y se limitan a la visualización de sus propios datos. 

Muchos usuarios han puesto manos a la obra, mediante técnicas como _web scrapping_ o elaboración propia han creado y compartido [Datasets sobre eSports](https://www.kaggle.com/search?q=eSports+in%3Adatasets ) que han sido de mucha utilidad para otros usuarios, sin embargo, la cantidad de DataSets que encontramos es limitada.

Esto es muy interesante ya que hay muchos usuarios interesados [[1]](https://www.reddit.com/r/esports/comments/74opfv/is_there_anywhere_i_can_find_a_dataset_in_regards/) [[2]](https://www.reddit.com/r/datasets/comments/4glujg/requestdatasets_about_the_esports_industry/) en DataSets. Ya sea para fines de análisis o inclusive académicos, se ven limitados por la poca disposición de Datasets listos para ser explotados. 

Pareciera que el acceso a esta información, quizás a causa de la _juventud_ de los eSports, se encuentra limitado al ecosistema que busca explotar estos datos
para aumentar el alcance e ingresos generados. 

# Antecedentes

[Leonardo Mauro](https://github.com/leomaurodesenv/) es un usuario de github y Data Scientist profesional que a través de la página [project-awesome](https://project-awesome.org/) nos comparte una lista de [_Datasets asombrosos_](https://project-awesome.org/leomaurodesenv/game-datasets) para cualquiera interesado en trabajar con **Inteligencia artificial** o **Data Mining** aplicado a videojuegos.

<p align="center">
<img src="https://project-awesome.org/images/logo.png" height ="250" width = "400" alt=API>
</p>

En esta página encontramos una gran colección de APIs, páginas web y DataSets, un trabajo _asombroso_ que debe ser reconocido. Es importante observar que no todos los Datasets están enfocados en videojuegos catalogados como eSports, y aquellos que los son, cuentan con pocos DataSets. 



# Propuesta: Aportar un Dataset para el análisis de eSports

Después de haber investigado sobre el panorama actual de los eSports, la importancia de estos y de la información que generan, así como las limitaciones con las que nos encontramos hoy en día para acceder a ella. Nos encontramos especialmente motivados para entrar al mar de información y tratar de aportar un granito de arena, un Dataset, que pueda ser de utilidad para todos los entusiastas de los eSports que estén interesados en su análisis.

## Referencias


1.   Hodge V., Devlin S., Sephton N., Block F., Drachen A., Cowling P. (2017). Win
prediction in eSports: mixed-rank match prediction in multi-player online battle arena games.
Recuperado de: https://arxiv.org/abs/1711.06498.
2.   Davenport, T.H. (2014). What businesses can learn from sports analytics. MIT Sloan, Management Review.
3.   Iqbal, M. Twitch revenues and usage statistics. BusinessOfApps, Febrero 2019. Recuperado de: www.businessofapps.com/data/twitch-statistics.
4.   Bozorgzadeh, A. (2017). Machine learning will be a game-changer for eSports. Venture Beat, Marzo 2017. Recuperado de: https://venturebeat.com/2017/03/11/machinelearning-will-be-a-game-changer-for-esports/.
5.   Lam, S. eSports as a goldmine for data analytics. ComputerWorldHK, Octubre 2017. Recuperado de: https://www.cw.com.hk/data-management/esports-as-a-goldmine-for-data-analytics.
6.   Joseph R, Travis F., Brett W. Northern Kentuchy University. E-Sports Analytics: A Primer and Resource for Student Research Projects and Lesson Plans. Journal of Instructional Pedagogies Volumen 23. Recuperado de: https://files.eric.ed.gov/fulltext/EJ1242004.pdf


# Postwork 02: Planteamiento de preguntas

Durante el módulo pasado nuestro [proyecto](https://github.com/LKXoro/BEDU-Rstudio-postworks/tree/main/Proyecto) estuvo relacionado con los eSports, de manera que pudimos responder algunas de las dudas que nos surgieron al inspeccionar [Datasets](https://github.com/LKXoro/BEDU-Rstudio-postworks/blob/main/Proyecto/Proyecto1/Readme.md) que obtuvimos en kaggle. La mayoría de estos se enfocan sustancialmente en dos aspectos de los eSports: **Gananacias y ganadores de torneos** (este último generalmente asociado a equipos profesionales).

También hicimos uso de un Dataset con información de los _streamers_ de Twitch que nos permitió conocer información como el promedio de visitas o número de seguidores de los canales oficiales de algunos eSports.

Nuestro proyecto anterior estaba enfocado a la evolución de los eSports en el tiempo, la escena actual y una posible predicción. Ahora que tenemos la oportunidad de revisar este tema nuevamente, buscamos concentrarnos otros aspectos como:

* eSports que se transmiten por internet
* El consumo de contenido relacionado a eSports
* Duración de las competencias
* Estado de salud de un videojuego considerado eSport

Es decir, queremos hablar de los juegos que hoy en día son eSports y responder preguntas como:

* Las partidas de eSports, ¿son enfrentamientos cortos o largos?
* ¿Cuántos torneos se han jugado de un juego en específico?
* ¿Cuántos torneos se han transmitido para que los fanáticos los vean?
* ¿Cuánto contenido relacionado a eSports consumen los fánaticos?
* ¿Los juegos que hoy en día son eSports los seguirán siendo en el futuro?

Estas y muchas otras preguntas pueden surgir si miramos el panorama actual, no desde una perspectiva de producto, sino de consumo. Como se muestra en este [artículo](https://contenidos.news/mediciones/nota/twitch-alcanzo-picos-de-73-millones-de-horas-de-consumo-diario-de-contenido)  latinoamérica es uno de los principales consumidores de contenido online relacionado a eSports a nivel mundial, una prueba más de la estrecha relación entre el creciente alcance de los eSports y el papel de los fanáticos en su popularidad.  

<p align="center">
<img src="https://contenidos.news/online/foto_189.jpg" height ="250" width = "400" alt=eSports>
</p>

Para poder responder a las preguntas realizadas, será necesario un dataset que nos brinde información de las competencias y torneos que contenga datos de tiempo, es posible que debamos buscar nuevas fuentes de información o atacar los Datasets que ya utilizamos desde otro ángulo. Como mencionamos en la identificación del problema, una de las mejores fuentes de datos son las APIs, así que son un buen punto de partida para encontrar la información que necesitamos y de ser necesaria adecuadala a nuestros fines.



# Postwork 03: Colección de datos
## PandaScore: Una API de eSports
<p align="center">
<img src="https://pandascore.co/assets/logo_for_twitter-0c1b471a256ce7364533695930bbe657ed3c8b2189f7ea07e60a40b614ebcf4e.png" alt=PandaScore_logo>
</p>

[PandaScore](https://pandascore.co/) provee estadísticas en tiempo real sobre eSports. Para lograr esto, recolecta información de enfrentamientos en competencias utilizando inteligencia artificial. 

Los datos que ofrece están disponibles en dos formas: 

*   Una [API REST](https://www.idento.es/blog/desarrollo-web/que-es-una-api-rest/) donde podemos encontrar toda la cobertura del calendario de eSports, así como estadísticas detallas antes y después de los enfrentamientos.
*   Una LIVE API: que nos brinda entrega un amplio rango de estadísticas en tiempo real mientras los enfrentamientos se llevan a cabo.

PandaScore ofrece distintos planes de pago, cada uno con mayor accesibilidad. El plan gratuito incluye acceso a la API REST donde podemos encontrar datos sobre los personajes de los juegos, calendarización de los enfrentamientos y datos sobre torneos, equipos y jugadores. Finalmente, es importante tener en cuenta que el plan gratuito permite realizar hasta 1000 peticiones por hora y podemos acceder a información sobre los siguientes juegos:

*   League of legends
*   CS:GO
*   Dota 2
*   Overwatch




# ¿Por qué seleccionamos PandaScore?

Seleccionamos PandaScore después de observar el contenido y tipo de datos que nos ofrece. La información que recibimos a más alto nivel es información de los distintos torneos que se han llevado a cabo, dentro podemos encontrar información sobre las diferentes ligas y los enfrentamientos que se dieron. También encontramos especialmente atractivos los campos relacionados con las fechas de inicio y fin, tanto de torneos como de enfrentamientos, así como un campo dedicado a especificar si el torneo o enfrentamiento fue transmitido por internet, y de ser así, información con el URL del stream en diferentes idiomas.


# ¿Cómo podemos realizar peticiones de datos a PandaScore?

Para lograr obtener los datos de PandaScore, primeramente debemos crear una cuenta de forma gratuita, que nos dará el acceso al plan básico. Una vez creada la cuenta se nos brindará un token o llave de acceso como la siguiente:



```py
token =  'V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns'
```


Para motivos de ejecución de nuestros códigos, y con la finalidad de que sea de utilidad para otros entusiastas del análisis de datos, de buena fé brindamos nuestro token de acceso para que, de ahora en adelante, podamos realizar las peticiones de datos a PandaScore.

Ya que este módulo esta enfocado al lenguaje [python](https://www.python.org/), recurrimos al paquete [request](https://realpython.com/python-requests/), idóneo para hacer peticiones http en python, ofreciendo instrucciones sencillas para enfocarnos en interactuar con los servicios y consumir los datos en nuestras aplicaciones.

<p align="center">
<img src="https://files.realpython.com/media/Python-Requests-Library-Tutorial_Watermarked.3c9dfdc7b014.jpg" height = '300' alt=requests>
</p>



## Formato de URL para peticiones a PandaScore

Antes de utilizar propiamente _request_, explicaremos el formato de URL que utilizaremos para ello. Como lo indica la [documentación](https://developers.pandascore.co/docs/tutorials/request_conditions) propia de PandaScore, la URL generalmente tiene el siguiente formato:

```py
url = "https://api.pandascore.co/[game]/[information]?page[number]=2&page[size]=50&token=[YOUR_TOKEN]"
```

Los parámetros que podemos modificar son los siguientes:


*   **[game]**: Dependiendo de la sintáxis de este parámetros se devuelve información de los diferentes juegos:
  * League of legends: 'lol'
  * Dota 2: 'dota2'
  * CS:GO: 'csgo'
  * Overwatch: 'overwatch'  
*   **[information]**: Acceso a información de:
  * Ligas: 'leagues' Son los eventos de mayor nivel, no tienen como tal una fecha y representan una competencia. Una liga se compone de varias _series_. Por ejemplo en deportes sería equivalente a la copa mundial de fútbol.
  * Series: 'series' Es una instancia de una liga, desde la fase de grupos hasta los _playoff_ finales. Ejemplo: Copa mundial de fútbol 2018
  * Torneos: 'tournaments' Los torneos son la manera en que los enfrentamientos son divividos en grupos para cada fase de la serie, habrá un torneo por grupo incluyendo la final. Ejemplo: Copa mundial de fútbol 2018 - Grupo C
  * Enfrentamientos: 'matches' Finalmente tenemos los enfrentamientos, entre dos oponentes y compuestos de varias partidas. Algunos enfrentamientos pueden reflejarse como "TBD vs TBD" cuando se conoce la fecha del enfrentamiento pero no al oponente del equipo. Ejemplo: Copa mundial de fútbol 2018 - Grupo C - Dinamarca vs Australia (1 juego)
  * Equipos: 'teams' Los nombres de los equipos que participan en estos eventos.
*   **Page[size]**: Este parámetro indica el número de resultados que se nos debe entregar en cada petición, el número de resultados máximos es de 100.
*   **Page[number]** Indica el número de página de la petición, entre menor sea el número de resultados, mayor será el número de páginas.
*   **[YOUR_TOKEN]**: Este parámetro debe ser sustituido por nuestra clave de acceso, la cual se nos brinda al crear una cuenta en PandaScore

Ahora bien, ya que conocemos la sintáxis adecuada para la URL, podemos llamar al paquete request y hacer uso de el:



In [ ]:
import requests #impotamos el paquete
#Parámetros deseados
juego = 'lol' #Queremos información de League of Legends
información = 'tournaments' #Los datos deben corresponder a torneos
registros = 100  #Queremos 100 registros por hoja
token =  'V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns'
# Aprovecharemos las bondades de los f string para introducir los parámetros
url = f"https://api.pandascore.co/{juego}/{información}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
#Por defecto si page[number] no es indicado, se nos devuelve la primer página
response = requests.get(url) #Realizamos la petición de información con el método get()
response.status_code #Verificamos el resultado de nuestra solicitud

200

Como podemos observar, obtuvimos una código de respuesta 200, ¡lo que indica que todo salió bien! la documentación también señala que dentro de la respuesta que obtuvimos podemos encontrar _headers_ con información útil como el total de registros que es posible obtener o el número de página en el que nos encontramos, a continuación de muestran los header de nuestra petición:

In [ ]:
print(response.headers['X-Page']) # Número de página
print(response.headers['X-Total']) # Total de registros
print(response.headers['X-Per-Page']) # Registros por página

1
824
100


Ahora para observar los datos que se nos devolvieron nos apoyaremos en el método json() y el paquete pandas, la información en _crudo_ que recibimos luce más o menos así:

In [ ]:
data = response.json()
data[0] # El método json devuelve una lista de diccionarios, entonces llamamos al primer elemento para ver su contenido
#Ya que la información que solicitamos es sbre torneos, la información devuelta es sobre el torneo más reciente

{'begin_at': '2021-03-09T17:00:00Z',
 'end_at': None,
 'id': 5649,
 'league': {'id': 4302,
  'image_url': 'https://cdn.pandascore.co/images/league/image/4302/220px-Prime_Leaguelogo_square.png',
  'modified_at': '2020-01-20T16:20:40Z',
  'name': 'Prime League Pro Division',
  'slug': 'league-of-legends-prime-league-pro-division',
  'url': 'https://www.primeleague.gg/de/start'},
 'league_id': 4302,
 'live_supported': False,
 'matches': [{'begin_at': '2021-03-09T17:00:00Z',
   'detailed_stats': True,
   'draw': False,
   'end_at': None,
   'forfeit': False,
   'game_advantage': None,
   'id': 586003,
   'live': {'opens_at': None, 'supported': False, 'url': None},
   'live_embed_url': None,
   'match_type': 'best_of',
   'modified_at': '2021-03-07T20:17:44Z',
   'name': 'Round 1 Match 1: GL vs USE',
   'number_of_games': 3,
   'official_stream_url': None,
   'original_scheduled_at': '2021-03-09T17:00:00Z',
   'rescheduled': False,
   'scheduled_at': '2021-03-09T17:00:00Z',
   'slug': 'team

¡oh rayos! esos son muchos campos, y más importante, tenemos listas dentro de los diccionarios. Antes de entrar en pánico intentemos ver como luce en un Dataframe de pandas:

In [ ]:
import pandas as pd #importamos el paquete pandas
df_torneos = pd.DataFrame.from_dict(data) #utilizamos el método Dataframe.from_dict() que es capaz de convertir listas de diccionarios a df
df_torneos.head() #Observemos las primeras filas del df

begin_at                end_at  ...  winner_id winner_type
0  2021-03-09T17:00:00Z                  None  ...        NaN        Team
1  2021-03-09T16:30:00Z                  None  ...        NaN        Team
2  2021-03-09T18:00:00Z  2021-03-23T23:00:00Z  ...        NaN        Team
3  2021-03-29T08:00:00Z  2021-04-09T14:00:00Z  ...        NaN        Team
4  2021-03-27T16:00:00Z  2021-04-18T21:00:00Z  ...        NaN        Team

[5 rows x 17 columns]

Nos encontramos con campos bastante amigables como las fechas de inicio y término de los torneos, el id del torneo, un campo para saber si fue transmitido en vivo, última actualización del registro, nombre del torneo y _slug_ (creemos hace referencia al nombre promocional del torneo, tenemos también el tipo de ganador del torneo y su id.  

Además, nos encontramos con cierta información en forma de listas dentro de campos, la cual es de difícil acceso y ciertamente algo confusa a la vista. Aunque quizás PandaScore ya nos haya brindado una solución a esto (uff).

Intentemos entonces solicitar a PandaScore esa información:

In [ ]:
información2 = 'leagues' #Cambiamos el tipo de información requerida

url2 = f"https://api.pandascore.co/{juego}/{información2}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"

response2 = requests.get(url2) 
response.status_code 

200

PandaScore ha respondido con un código 200, veamos que nos devolvió:

In [ ]:
print(response2.headers['X-Page']) # Número de página
print(response2.headers['X-Total']) # Total de registros
print(response2.headers['X-Per-Page']) # Registros por página

1
67
100


Tenemos en nuestras manos la primer página, y aparentemente la única al haber solicitado 100 registros y tener en total 67. Veamos el primer elemento:

In [ ]:
data2 = response2.json()
data2[0] 

{'id': 4539,
 'image_url': 'https://cdn.pandascore.co/images/league/image/4539/LCO_logo.png',
 'modified_at': '2021-02-18T10:32:45Z',
 'name': 'LCO',
 'series': [{'begin_at': '2021-02-23T07:00:00Z',
   'description': None,
   'end_at': '2021-04-09T22:00:00Z',
   'full_name': 'Split 1 2021',
   'id': 3361,
   'league_id': 4539,
   'modified_at': '2021-02-18T11:40:54Z',
   'name': 'Split 1',
   'season': None,
   'slug': 'league-of-legends-lco-split-1-2021',
   'tier': 'c',
   'winner_id': None,
   'winner_type': None,
   'year': 2021}],
 'slug': 'league-of-legends-lco',
 'url': 'https://lco.gg/',
 'videogame': {'current_version': '11.5.1',
  'id': 1,
  'name': 'LoL',
  'slug': 'league-of-legends'}}

Tenemos un diccionario con menos campos, exploremos más a fondo con el dataframe:

In [ ]:
df_ligas = pd.DataFrame.from_dict(data2) #utilizamos el método Dataframe.from_dict() que es capaz de convertir listas de diccionarios a df
df_ligas.head() #Observemos las primeras filas del df

id  ...                                          videogame
0  4539  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
1  4533  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
2  4483  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
3  4450  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
4  4433  ...  {'current_version': '11.5.1', 'id': 1, 'name':...

[5 rows x 8 columns]

Inspeccionando un poco nos encontramos que los datos devueltos en una solicitud de _ligas_ es diferentes al campo _liga_ de los torneos, es bueno saber esto para evitar confusiones. Antes de pensar en las posibilidades, veamos las 3 solicitudes restantes que podemos observar:

In [ ]:
información3 = 'series' 
información4 = 'matches'
información5 = 'teams'

url3 = f"https://api.pandascore.co/{juego}/{información3}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url4 = f"https://api.pandascore.co/{juego}/{información4}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url5 = f"https://api.pandascore.co/{juego}/{información5}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
 
response3 = requests.get(url3) 
response4 = requests.get(url4) 
response5 = requests.get(url5) 

print(response3.status_code)
print(response4.status_code)
print(response5.status_code)


200
200
200


Parece que todo salió bien con las peticiones, veamos los dataframes:

In [ ]:
data3 = response3.json()
data4 = response4.json()
data5 = response5.json()

df_series = pd.DataFrame.from_dict(data3)
df_enfrentamientos = pd.DataFrame.from_dict(data4)
df_equipos = pd.DataFrame.from_dict(data5)

In [ ]:
df_series.head()

begin_at description  ... winner_type  year
0  2021-02-23T07:00:00Z        None  ...        None  2021
1  2021-01-25T18:00:00Z        None  ...        None  2021
2  2021-02-13T13:00:00Z        None  ...        None  2021
3  2021-02-19T08:00:00Z        None  ...        None  2021
4  2021-01-19T19:00:00Z        None  ...        None  2021

[5 rows x 18 columns]

In [ ]:
df_enfrentamientos.head()

forfeit  ...                                               live
0    False  ...  {'opens_at': None, 'supported': False, 'url': ...
1    False  ...  {'opens_at': None, 'supported': False, 'url': ...
2    False  ...  {'opens_at': None, 'supported': False, 'url': ...
3    False  ...  {'opens_at': None, 'supported': False, 'url': ...
4    False  ...  {'opens_at': None, 'supported': False, 'url': ...

[5 rows x 33 columns]

In [ ]:
df_equipos.head()

acronym  ...                        slug
0     PCE  ...                       peace
1     Z10  ...                  z10-eshark
2     X25  ...                 x25-esports
3    None  ...                   ctrl-play
4     IPN  ...  impunity-league-of-legends

[5 rows x 9 columns]

¡Increíble! Con tan solo 5 dataframes tenemos muchísima información acerca de la escena competitiva de League of Legends, pero si recordamos bien... ¡tenemos acceso a otros 3 juegos! entonces en total tendremos 20 dataframes con los que podríamos trabajar ( _panic_ ). De momentos nos enfocaremos en trabajar con solo los dataframes obtenidos para LoL, quizás haya una forma de trabajar con todos, una luz al final del túnel.

# Postwork 04: Análisis Exploratorio de nuestro Dataset

En el [postwork anterior](https://github.com/LKXoro/BEDU-Rstudio-postworks/blob/main/M3_Postwork3/Postwork_03_python.ipynb) exploramos las posibles respuestas que nos brinda PandaScore. Con ellas obtuvimos 5 Datasets llenos de información tan sólo para League of Legends. Ahora vamos a inspeccionar un poco los campos que tenemos y el tipo de información dentro de ellos, comencemos realizando las peticiones a la API:

In [ ]:
import requests 

juego = 'lol' 
información = 'tournaments' 
información2 = 'leagues'
información3 = 'series' 
información4 = 'matches'
información5 = 'teams'
registros = 100  
token =  'V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns'

url = f"https://api.pandascore.co/{juego}/{información}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url2 = f"https://api.pandascore.co/{juego}/{información2}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url3 = f"https://api.pandascore.co/{juego}/{información3}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url4 = f"https://api.pandascore.co/{juego}/{información4}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url5 = f"https://api.pandascore.co/{juego}/{información5}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"

response = requests.get(url) 
response2 = requests.get(url2)
response3 = requests.get(url3) 
response4 = requests.get(url4) 
response5 = requests.get(url5) 

print(response.status_code)
print(response2.status_code)
print(response3.status_code)
print(response4.status_code)
print(response5.status_code)


200
200
200
200
200


¡Genial! Todas las peticiones se realizaron correctamente. Vamos a explorar un poco la cantidad de datos que se nos devolvió:

In [ ]:
print(f'### La petición de {información} nos devolvió:\n')
print(response.headers['X-Total']+' registros totales.\n') 
print(f'### La petición de {información2} nos devolvió:\n')
print(response2.headers['X-Total']+' registros totales.\n') 
print(f'### La petición de {información3} nos devolvió:\n')
print(response3.headers['X-Total']+' registros totales.\n') 
print(f'### La petición de {información4} nos devolvió:\n')
print(response4.headers['X-Total']+' registros totales.\n') 
print(f'### La petición de {información5} nos devolvió:\n')
print(response5.headers['X-Total']+' registros totales.\n')   

### La petición de tournaments nos devolvió:

824 registros totales.

### La petición de leagues nos devolvió:

67 registros totales.

### La petición de series nos devolvió:

298 registros totales.

### La petición de matches nos devolvió:

18694 registros totales.

### La petición de teams nos devolvió:

973 registros totales.



Observamos que los datos relacions con los enfrentamientos (matches) cuentan con una extensa cantidad de registros, esto es lógico, ya que dentro de cada torneo ha habido diversos enfrentamientos (¡Y han habido muchos torneos!). Exploremos mejor los campos de cada respuesta, primero creamos los df:

In [ ]:
import pandas as pd

data1 = response.json()
data2 = response2.json()
data3 = response3.json()
data4 = response4.json()
data5 = response5.json()

df_torneos = pd.DataFrame.from_dict(data1)
df_ligas= pd.DataFrame.from_dict(data2)
df_series = pd.DataFrame.from_dict(data3)
df_enfrentamientos = pd.DataFrame.from_dict(data4)
df_equipos = pd.DataFrame.from_dict(data5)

Si seguimos la jerarquía de organización de los eventos de eSports, el primer elemento que debemos inspeccionar son las ligas:

In [ ]:
df_ligas.head()

id  ...                                          videogame
0  4539  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
1  4533  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
2  4483  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
3  4450  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
4  4433  ...  {'current_version': '11.5.1', 'id': 1, 'name':...

[5 rows x 8 columns]

Nos encontramos ante un df con 8 campos, dos de ellos son url que nos llevan a la página oficial (en caso de tenerla) y el logo oficial de la serie. También tenemos un campo para la versión del juego con la que se llevo a cabo la liga y las distintas series que se han llevado a cabo, el slug (¿eslogan?) de la liga y el nombre de la liga.

La cantidad de datos sobre ligas, al ser los eventos de mayor nivel, es la de menor cantidad. El número de campos es algo limitado y no todos son de útilidad para análisis, para un dashboard los url e imágenes serían ideales.

Veamos que información tenemos respecto a las series:

In [ ]:
df_series.head()

begin_at description  ... winner_type  year
0  2021-02-23T07:00:00Z        None  ...        None  2021
1  2021-01-25T18:00:00Z        None  ...        None  2021
2  2021-02-13T13:00:00Z        None  ...        None  2021
3  2021-02-19T08:00:00Z        None  ...        None  2021
4  2021-01-19T19:00:00Z        None  ...        None  2021

[5 rows x 18 columns]

En este Dataframe encontramos más campos como la fecha de inicio y fin de la serie, un campo de descripción que parece estar lleno de valores None al igual que el nombre del juego, el nombre de la serie y slug, la temporada y nivel y se la serie,encontramos también el id de la liga correspondiente y el id del ganador (si no se ha definido un ganador se tiene un valor NaN), el tipo de ganador (LoL es un juego de equipos así que en general el tipo de ganador es Team y también tenemos el año al que corresponde la serie.

Dentro de estos campos también tenemos el que corresponde al de los torneos, veamos que tenemos respecto a ellos:

In [ ]:
df_torneos.head()

begin_at                end_at  ...  winner_id winner_type
0  2021-03-09T17:00:00Z                  None  ...        NaN        Team
1  2021-03-09T16:30:00Z                  None  ...        NaN        Team
2  2021-03-09T18:00:00Z  2021-03-23T23:00:00Z  ...        NaN        Team
3  2021-03-29T08:00:00Z  2021-04-09T14:00:00Z  ...        NaN        Team
4  2021-03-27T16:00:00Z  2021-04-18T21:00:00Z  ...        NaN        Team

[5 rows x 17 columns]

Encontramos las fechas de inicio y fin de cada torneo, su id de torneo junto a los id de la liga y serie a la que corresponde. Fecha de última actualización, soporte de stream en vivo, nombre del torneo, slug y prizepool que parece no tener un valor diferente a None, tenemos el winner id y winner type (en caso de haber un ganador) y dos campos dedicados a los enfrentamientos y equipos.

En los enfrentamientos encontramos aún más información:

In [ ]:
df_enfrentamientos.head()

name  ...              begin_at
0            Round 3: TBD vs TBD  ...  2021-03-11T17:00:00Z
1   Round 2 Match 2: MOUZ vs BIG  ...  2021-03-10T20:00:00Z
2    Round 2 Match 1: TBD vs TBD  ...  2021-03-10T17:00:00Z
3  Round 1 Match 2: SKP vs S04.E  ...  2021-03-09T20:00:00Z
4     Round 1 Match 1: GL vs USE  ...  2021-03-09T17:00:00Z

[5 rows x 33 columns]


Parece que este Dataframe contiene información que nos puede ayudar a analizar el performance de los equipos y además tenemos datos sobre las partidas y sus streams. Encontramos las fechas de inicio y fin de los enfrentamientos, los id de ligas y torneos, resultados, la versión del juego con la que se jugó, tipos de enfrentamiento, número de partidas por enfrentamiento, si hubo abandonó o cambios de calendarización, etc.

Hasta el momento los datos respecto a los enfrentamientos parecen aportar más detalles. Finalmente, echemos un ojo a los equipos:

In [ ]:
df_equipos.head()

acronym  ...                        slug
0     PCE  ...                       peace
1     Z10  ...                  z10-eshark
2     X25  ...                 x25-esports
3    None  ...                   ctrl-play
4     IPN  ...  impunity-league-of-legends

[5 rows x 9 columns]

Al igual que el Dataframe de las ligas, los datos de los equipos parecen estar más limitados a sus nombres, logotipos promocionales, páginas web y algún que otro dato de sus integrantes. Después de está inspección preliminar parece claro que la información más jugosa la encontramos directamente en el calor de la batalla, es decir, los enfrentamientos.

Inspeccionemos sus tipos de datos:

In [ ]:
df_enfrentamientos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 33 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   name                   100 non-null    object 
 1   official_stream_url    93 non-null     object 
 2   end_at                 37 non-null     object 
 3   draw                   100 non-null    bool   
 4   league                 100 non-null    object 
 5   league_id              100 non-null    int64  
 6   detailed_stats         100 non-null    bool   
 7   live_embed_url         93 non-null     object 
 8   tournament_id          100 non-null    int64  
 9   results                100 non-null    object 
 10  videogame_version      37 non-null     object 
 11  streams                100 non-null    object 
 12  modified_at            100 non-null    object 
 13  games                  100 non-null    object 
 14  original_scheduled_at  100 non-null    object 
 15  match_t

Recordemos que este Dataframe es solo una pequeña parte de todos los datos que nos puede entregar la API, 100 registros de 18694 en realidad, pero podemos sacar algunas conclusiones preliminares.

Es necesario transformar los campos de fechas al tipo de dato correspondiente para su manejo, tal vez sea necesario eliminar algunas columnas que sea redundantes o que no aporten información útil, como game advantage que parece estar lleno de valores None. Puede que sea conveniente transformar algunos datos a tipo string para obtener más información.

Una de las cosas que pueden resultar problemáticas son los enfrentamientos que estan agendados pero no se han llevado a cabo, si bien es bueno tenerlos en cuenta, muchos de los campos terminan teniendo valores nulos a causa de estos, por lo que resulta conveniente pensar solo en aquellos enfrentamientos que ya se han llevado a cabo.

Hay muchas posibilidades, pero por el momento, parece que el dataframe que mejor podemos explotar para responder a las preguntas que nos hicimos es el correspondiente a los enfrentamientos.

# Postworks 05 y 06: Limpieza de datos y agregaciones / Automatización y APIs

En esencia el Postwork 05 busca que limpiemos nuestro Dataset de valores nulos y probemos obtener información mediante agreaciones. Mientras que el Postwork 06 invita a automatizar el proceso de peticiones a una API. En esta entrega cambiaremos un poco el orden, primero realizaremos la automatización de la petición para obtener todos los datos y después los limpiaremos.

<p align="center">
<img src="https://empresas.blogthinkbig.com/wp-content/uploads/2019/03/janitor.png" height ="200" width = "200" alt=lego>
</p>

En postworks anteriores mencionamos que una petición a nuestra API de confianza, PandaScore, puede regresar como máximo 100 registros por petición y que las peticiones responden a parámetros de la URL. Creemos que la mejor forma de automatizar este proceso es con una función que nos devuelva la información que queremos, del juego que queremos de la API:

In [ ]:
import requests #Importamos los paquete necesarios
import pandas as pd 
import math as mt
 
def PandaScore_get(juego,informacion): #Definimos nuestra función
    if isinstance(juego, str) and isinstance(informacion, str): #Nos aseguramos que los argumentos sean de tipo str
        registros = 100 #Establecemos el número de registros por página
        token = 'V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns' #Token de acceso
        url = f"https://api.pandascore.co/{juego}/{informacion}?page[size]={registros}&token={token}" #URL con los parámetros deseados
        response = requests.get(url) #Realizamos la petición
        if response.status_code == 200: #Si todo salió bien continuamos
            Iteraciones = mt.ceil(float(response.headers['X-Total'])/registros) #Calculamos el número de páginas en base a todos los registros que hay en la API
            #Al dividir el número total de registros entre los registros solicitados obtenemos un aproximado, para asegurar que tomaremos todos, redondeamos al entero más alto
            for i in range(1,Iteraciones+1): #Ciclo for para obtener todos los registros de cada página
                url = f"https://api.pandascore.co/{juego}/{informacion}?page[size]={registros}&page[number]={i}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
                #URL el parámetro de página deseada
                response = requests.get(url) #Realizamos la petición correspondiente
                data = response.json() #Transformamos la respuesta 
                if i == 1: #En la primer iteración se crea un Dataframe
                    df = pd.DataFrame.from_dict(data)
                else: #En el resto de iteraciones se van añadiendo los nuevos registros
                    df = df.append(pd.DataFrame.from_dict(data),ignore_index=True) 
            return df # Finalmente se devuelve el Dataframe con todos los registros
        else: #Si la petición no se ejecutó correctamente mandamos un mensaje con el error devuelto
            return 'Error de petición: '+str(response.status_code)  
    else: #Si los argumentos no son de tipo str regresamos un mensaje
        return 'Por favor coloque un argumento válido en formato string.'
  

Veamos a la función en acción:

In [ ]:
PandaScore_get('lol',1)

'Por favor coloque un argumento válido en formato string.'

In [ ]:
PandaScore_get('lol','matchs')

'Error de petición: 404'

Si los argumentos son correctos se nos devolverá el Dataframe correspondiente, el tiempo de ejecución varia dependiendo de la información solicitada, en el caso de los enfrentamientos tenemos 18694 registros, por lo que la función tomará un par de minutos al solo poder realizar 100 peticiones, en general consumiriamos 187 de nuestros 1000 intentos por hora:

In [ ]:
df_matches = PandaScore_get('lol','matches')

Observemos el Dataframe que se nos devuelve:

In [ ]:
df_matches

name  ...              begin_at
0                Round 3: TBD vs TBD  ...  2021-03-11T17:00:00Z
1       Round 2 Match 2: MOUZ vs BIG  ...  2021-03-10T20:00:00Z
2        Round 2 Match 1: TBD vs TBD  ...  2021-03-10T17:00:00Z
3      Round 1 Match 2: SKP vs S04.E  ...  2021-03-09T20:00:00Z
4         Round 1 Match 1: GL vs USE  ...  2021-03-09T17:00:00Z
...                              ...  ...                   ...
18689                      DIG vs C9  ...  2015-05-31T20:09:51Z
18690                       T8 vs TL  ...  2015-05-31T18:22:58Z
18691                      GV vs NME  ...  2015-05-30T23:23:25Z
18692                      T8 vs TIP  ...  2015-05-30T22:23:32Z
18693                      TL vs TDK  ...  2015-05-30T20:11:29Z

[18694 rows x 33 columns]

Veamos las columnas y tipos de datos:

In [ ]:
df_matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18694 entries, 0 to 18693
Data columns (total 33 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   name                   18694 non-null  object 
 1   official_stream_url    10896 non-null  object 
 2   end_at                 18135 non-null  object 
 3   draw                   18694 non-null  bool   
 4   league                 18694 non-null  object 
 5   league_id              18694 non-null  int64  
 6   detailed_stats         18694 non-null  bool   
 7   live_embed_url         9340 non-null   object 
 8   tournament_id          18694 non-null  int64  
 9   results                18694 non-null  object 
 10  videogame_version      17824 non-null  object 
 11  streams                18694 non-null  object 
 12  modified_at            18694 non-null  object 
 13  games                  18694 non-null  object 
 14  original_scheduled_at  15786 non-null  object 
 15  ma

Como podemos ver tenemos 33 columnas, un número considerable, observemos que contienen algunas y si nos son de utilidad. Empecemos por los campos que contienen listas:

In [ ]:
df_matches['videogame'][0] 

{'id': 1, 'name': 'LoL', 'slug': 'league-of-legends'}

El campo _videogame_ puede que haya sido de utilidad para la API al momento de regresarnos una respuesta, sin embargo, es  redundante para nosotros.

In [ ]:
df_matches['slug']

0          2021-03-11-c3b8c131-fac1-4b71-b339-f1d0387914a6
1        mousesports-vs-berlin-international-gaming-202...
2          2021-03-10-7412caf0-3f45-40e1-b763-80d005a04162
3        sk-gaming-prime-vs-fc-schalke-04-evolution-202...
4        team-gamerlegion-vs-unicorns-of-love-sexy-edit...
                               ...                        
18689                        dignitas-vs-cloud9-2015-05-31
18690                     team-8-vs-team-liquid-2015-05-31
18691                          gravity-vs-enemy-2015-05-31
18692                         team-8-vs-impulse-2015-05-31
18693             team-liquid-vs-dragon-knights-2015-05-30
Name: slug, Length: 18694, dtype: object

Una combinación del campo nombre y la fecha programada. Además de que en algunos casos tenemos valores raros, consideramos que podemos retirar esta columna.

In [ ]:
df_matches['streams'][0]

{'english': {'embed_url': None, 'raw_url': None},
 'official': {'embed_url': None, 'raw_url': None},
 'russian': {'embed_url': None, 'raw_url': None}}

In [ ]:
df_matches['streams'][300]

{'english': {'embed_url': None, 'raw_url': None},
 'official': {'embed_url': 'https://player.twitch.tv/?channel=riotgamesru',
  'raw_url': 'https://www.twitch.tv/riotgamesru'},
 'russian': {'embed_url': 'https://player.twitch.tv/?channel=riotgamesru',
  'raw_url': 'https://www.twitch.tv/riotgamesru'}}

In [ ]:
df_matches.loc[300,'official_stream_url']

'https://www.twitch.tv/riotgamesru'

In [ ]:
df_matches['live_embed_url'][300]

'https://player.twitch.tv/?channel=riotgamesru'

In [ ]:
df_matches['official_stream_url'][300]

'https://www.twitch.tv/riotgamesru'

El campo _streams_ nos regresa las direcciones URL de las transmisiones en vivo en diferentes idiomas, mientras que el campo _live embed url_ nos da el enlace en el que se transmitió en vivo y _official stream url_ nos regresa únicamente la URL del canal oficial que realizo el streaming. Creemos es mejor conservar este último campo con la dirección oficial a un campo con una lista  más difícil de manejar y otro con URLs que ya no son accesibles.

In [ ]:
df_matches['tournament'][0]

{'begin_at': '2021-03-09T17:00:00Z',
 'end_at': None,
 'id': 5649,
 'league_id': 4302,
 'live_supported': False,
 'modified_at': '2021-03-08T14:50:51Z',
 'name': 'Playoffs',
 'prizepool': None,
 'serie_id': 3254,
 'slug': 'league-of-legends-prime-league-pro-division-spring-2021-playoffs',
 'winner_id': None,
 'winner_type': 'Team'}

El campo _tournament_ nos devuelve el torneo correspondiente. Sin embargo, contamos con el _tournament id_, y podemos realizar peticiones de torneos si lo deseamos, por lo que de ser necesario podriamos realizar un **join** y observar a que torneo corresponde cada enfrentamiento. 

In [ ]:
df_matches['modified_at'].head() 

0    2021-03-07T20:20:32Z
1    2021-03-07T20:19:53Z
2    2021-03-07T20:19:33Z
3    2021-03-07T20:18:12Z
4    2021-03-07T20:17:44Z
Name: modified_at, dtype: object

La última fecha de modificación respecto a la API no nos resulta útil del todo a nosotros.

In [ ]:
df_matches['league'][0]

{'id': 4302,
 'image_url': 'https://cdn.pandascore.co/images/league/image/4302/220px-Prime_Leaguelogo_square.png',
 'modified_at': '2020-01-20T16:20:40Z',
 'name': 'Prime League Pro Division',
 'slug': 'league-of-legends-prime-league-pro-division',
 'url': 'https://www.primeleague.gg/de/start'}

In [ ]:
df_matches['serie'][0]

{'begin_at': '2021-01-19T17:00:00Z',
 'description': None,
 'end_at': None,
 'full_name': 'Spring 2021',
 'id': 3254,
 'league_id': 4302,
 'modified_at': '2021-01-12T08:59:18Z',
 'name': '',
 'season': 'Spring',
 'slug': 'league-of-legends-prime-league-pro-division-spring-2021',
 'tier': 'c',
 'winner_id': None,
 'winner_type': None,
 'year': 2021}

Mismo caso que con los torneos, si conservamos el _league id_ y _serie id_ no deberiamos tener problemas.

In [ ]:
df_matches['id']

0        586007
1        586006
2        586005
3        586004
4        586003
          ...  
18689         7
18690         6
18691         5
18692         4
18693         2
Name: id, Length: 18694, dtype: int64

Id asignado por la API a cada enfrentamiento, con el observamos que los enfrentamientos en los primeros registros son los más recientes y los últmos son los más antiguos. Para evitar confusiones y por el tratamiento que le daremos a los datos será mejor usar nuestra propia indexación.

In [ ]:
df_matches['games'][300]

[{'begin_at': None,
  'complete': False,
  'detailed_stats': True,
  'end_at': None,
  'finished': False,
  'forfeit': False,
  'id': 222685,
  'length': None,
  'match_id': 583283,
  'position': 1,
  'status': 'not_started',
  'video_url': None,
  'winner': {'id': None, 'type': None},
  'winner_type': None}]

El campo _games_ debería entregar información de cada partida en el enfrentamiento, pero en general esos datos podemos encontrarlos en otros campos que ya tenemos.

In [ ]:
df_matches['game_advantage'].unique()

array([None, nan, 2862.0, 318.0, 126059.0, 329.0, 74.0, 1566.0, 249.0,
       205.0, 411.0, 65.0, 267.0], dtype=object)


El campo game_advantage es poco claro, en la documentación no se específica el significado de estos valores.

In [ ]:
df_matches['original_scheduled_at']

0        2021-03-11T17:00:00Z
1        2021-03-10T20:00:00Z
2        2021-03-10T17:00:00Z
3        2021-03-09T20:00:00Z
4        2021-03-09T17:00:00Z
                 ...         
18689                    None
18690                    None
18691                    None
18692                    None
18693                    None
Name: original_scheduled_at, Length: 18694, dtype: object

In [ ]:
df_matches['scheduled_at']

0        2021-03-11T17:00:00Z
1        2021-03-10T20:00:00Z
2        2021-03-10T17:00:00Z
3        2021-03-09T20:00:00Z
4        2021-03-09T17:00:00Z
                 ...         
18689    2015-05-31T20:00:00Z
18690    2015-05-31T18:00:00Z
18691    2015-05-30T23:00:00Z
18692    2015-05-30T22:00:00Z
18693    2015-05-30T20:00:00Z
Name: scheduled_at, Length: 18694, dtype: object

Estos campos pueden indicar modificaciones en las fechas de los enfrentamientos, pero el campo _rescheduled_ nos brinda la misma información de forma más concisa.

In [ ]:
df_matches['match_type'].unique()

array(['best_of'], dtype=object)

En general todos los enfrentamientos son del tipo "_el mejor de_", así que este campo también es redundante.

In [ ]:
df_matches['detailed_stats']

0        True
1        True
2        True
3        True
4        True
         ... 
18689    True
18690    True
18691    True
18692    True
18693    True
Name: detailed_stats, Length: 18694, dtype: bool

Un campo que indica si el campo _games_ contiene detalles, pero ya que no vamos a utilizar _games_ podemos omitirlo.

In [ ]:
df_matches['videogame_version']

0                                       None
1                                       None
2                                       None
3                                       None
4                                       None
                        ...                 
18689    {'current': False, 'name': '5.8.1'}
18690    {'current': False, 'name': '5.8.1'}
18691    {'current': False, 'name': '5.8.1'}
18692    {'current': False, 'name': '5.8.1'}
18693    {'current': False, 'name': '5.9.1'}
Name: videogame_version, Length: 18694, dtype: object

Un campo que podría ser de interés si analizamos las composiciones de _personajes_ de cada equipo respecto a la versión del juego, lamentablemente no contamos con esa información.

In [ ]:
df_matches['live'][0]

{'opens_at': None, 'supported': False, 'url': None}

Finalmente el campo _live_ nos brinda la url del stream si el enfrentamiento está en vivo en ese momento, en general si buscaramos un soporte en vivo sería un campo genial, pero como deseamos trabajar con resultados de enfrentamientos que ya han terminado podemos trabajar sin este campo.

Ahora eliminamos los campos que deseamos retirar:

In [ ]:
campos_no_deseados = ['videogame', 'slug', 'tournament', 'league', 'serie', 'streams', 'modified_at', 'games',
                      'game_advantage', 'scheduled_at', 'original_scheduled_at', 'detailed_stats','live',
                      'match_type', 'videogame_version', 'live_embed_url', 'id']

df_matches_temp = df_matches.drop(columns=campos_no_deseados)
df_matches_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18694 entries, 0 to 18693
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   name                 18694 non-null  object 
 1   official_stream_url  10896 non-null  object 
 2   end_at               18135 non-null  object 
 3   draw                 18694 non-null  bool   
 4   league_id            18694 non-null  int64  
 5   tournament_id        18694 non-null  int64  
 6   results              18694 non-null  object 
 7   winner               17770 non-null  object 
 8   status               18694 non-null  object 
 9   winner_id            17770 non-null  float64
 10  forfeit              18694 non-null  bool   
 11  rescheduled          18694 non-null  bool   
 12  serie_id             18694 non-null  int64  
 13  opponents            18694 non-null  object 
 14  number_of_games      18694 non-null  int64  
 15  begin_at             18662 non-null 

In [ ]:
df_matches_temp.head()

name  ...              begin_at
0            Round 3: TBD vs TBD  ...  2021-03-11T17:00:00Z
1   Round 2 Match 2: MOUZ vs BIG  ...  2021-03-10T20:00:00Z
2    Round 2 Match 1: TBD vs TBD  ...  2021-03-10T17:00:00Z
3  Round 1 Match 2: SKP vs S04.E  ...  2021-03-09T20:00:00Z
4     Round 1 Match 1: GL vs USE  ...  2021-03-09T17:00:00Z

[5 rows x 16 columns]

¡Genial¡, pudimos reducir el número de campos de 33 a 16, ahora veamos de que forma podemos tratar con los valores _None_. Si inspeccionamos por columna:

In [ ]:
df_matches_temp.isna().sum(axis=0)

name                      0
official_stream_url    7798
end_at                  559
draw                      0
league_id                 0
tournament_id             0
results                   0
winner                  924
status                    0
winner_id               924
forfeit                   0
rescheduled               0
serie_id                  0
opponents                 0
number_of_games           0
begin_at                 32
dtype: int64

Observamos que la mayor cantidad de valores nulos se da en _official stream url_ esto no es del todo un problema ya que no necesariamente todos los partidos se transmitieron en directo. Entonces nos encontramos con valores nulos en las fechas de inicio y terminó, seguramente por partidos que todavía no se han llevado a cabo, se están llevando a cabo en estos momentos o se cancelaron, por ende también el valor del campo _winner_ tiene un valor también nulo.

Aprovechemos el campo status para limpiar nuestro dataframe:

In [ ]:
df_matches_temp['status'].unique()

array(['not_started', 'finished', 'canceled'], dtype=object)

Obtengamos solo aquellos registros donde el estado del enfrentamiento sea "_finished_":

In [ ]:
df_matches_finished = df_matches_temp[ df_matches_temp['status'] == 'finished' ]
df_matches_finished.isna().sum(axis=0)

name                      0
official_stream_url    7772
end_at                   19
draw                      0
league_id                 0
tournament_id             0
results                   0
winner                  398
status                    0
winner_id               398
forfeit                   0
rescheduled               0
serie_id                  0
opponents                 0
number_of_games           0
begin_at                  1
dtype: int64

Aún conservamos un par de valores None en nuestras fechas, esto puede ser a causa de errores en la recopilación de los datos. Deshagámonos de ellos:

In [ ]:
df_matches_finished_clean = df_matches_finished.dropna(subset=['begin_at', 'end_at'])
df_matches_finished_clean.isna().sum(axis=0)

name                      0
official_stream_url    7755
end_at                    0
draw                      0
league_id                 0
tournament_id             0
results                   0
winner                  398
status                    0
winner_id               398
forfeit                   0
rescheduled               0
serie_id                  0
opponents                 0
number_of_games           0
begin_at                  0
dtype: int64

¡Enhorabuena! tenemos un dataframe solo con las partidas que han sido finalizadas y además tienen una fecha de fin e inicio establecidas. ¿Pero qué pasa con esos 398 valores _None_ en el campo _winner id_ y _winner_? ¿Y los 7735 de _official stream url_? Pues tienen una explicación muy sencilla en realidad. Estos datos quieren decir que 398 enfrentamiento han terminado en empate y 7735 enfrentamientos no se transmitieron en línea.

Para dejar nuestro dataframe un poco más presentable vamos a ordenar los campos de formas más amigable:

In [ ]:
#No incluiremos al campo status por que ya ha cumplido su propósito
df_matches_ordenado = df_matches_finished_clean[['name','opponents','results','forfeit','draw',
                                                      'number_of_games','winner','winner_id','begin_at','end_at',
                                                      'rescheduled','official_stream_url','league_id', 'serie_id', 'tournament_id']]

Observemos al nuevo Dataframe:

In [ ]:
df_matches_ordenado.head()

name  ... tournament_id
15                                   ALF vs BJD  ...          5483
16    Lower Bracket Round 2 Match 1: CGG vs GGE  ...          5579
36                  Round 1 match 1: KVM vs ION  ...          5609
37  Winners bracket round 1 match 2: ESB vs SIN  ...          5602
38   Winners bracket round 1 match 1: IG vs VIK  ...          5602

[5 rows x 15 columns]

Reiniciemos el índice y cambiemos algunos nombre de los campos para obtener un Dataframe más limpio:

In [ ]:
df_matches_ordenado = df_matches_ordenado.reset_index(drop=True) 
nombres_nuevos = {
    'name' : 'match_name',
    'begin_at' : 'begin_date',
    'end_at' : 'end_date',
    'official_stream_url' : 'official_stream_channel_url'
}

df_matches_ordenado.rename(columns=nombres_nuevos, inplace=True)
df_matches_ordenado.head()

match_name  ... tournament_id
0                                   ALF vs BJD  ...          5483
1    Lower Bracket Round 2 Match 1: CGG vs GGE  ...          5579
2                  Round 1 match 1: KVM vs ION  ...          5609
3  Winners bracket round 1 match 2: ESB vs SIN  ...          5602
4   Winners bracket round 1 match 1: IG vs VIK  ...          5602

[5 rows x 15 columns]

¡Bien! Tenemos un Dataframe más limpio y ordenado, con ayuda de algunos métodos de agregación, podemos aprovechar los valores _none_ que hemos dejado, por ejemplo, para conocer la proporción de enfrentamientos transmitidos:

In [ ]:
sin_stream = df_matches_ordenado['official_stream_channel_url'].isna().sum()
total = len(df_matches_ordenado['official_stream_channel_url'])

proporción = (sin_stream/total)*100
proporción 

42.80746301611835

Ahora sabemos que aproximadamente el 42.83% de los enfrentamientos de lol son transmitidos en línea. Finalmente guardemos nuestro Dataframe en un archivo csv:

In [ ]:
df_matches_ordenado.to_csv('lol_matches.csv')

# Postwork 07: Transformación, filtración y ordenamiento de datos 
Antes de comenzar en caso de no tener nuestro confiable archivo csv a la mano, o bien debido a que la API está en constante actualización podemos recurrir a una versión mejorada de la función para peticiones de los **enfrentamientos**:

In [ ]:
import requests #Importamos los paquete necesarios
import pandas as pd 
import math as mt
 
def PandaScore_get_matches(juego): #Definimos nuestra función
    if isinstance(juego, str): #Nos aseguramos que los argumentos sean de tipo str
        informacion = 'matches'
        registros = 100 #Establecemos el número de registros por página
        token = 'V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns' #Token de acceso
        url = f"https://api.pandascore.co/{juego}/{informacion}?page[size]={registros}&token={token}" #URL con los parámetros deseados
        response = requests.get(url) #Realizamos la petición
        if response.status_code == 200: #Si todo salió bien continuamos
            Iteraciones = mt.ceil(float(response.headers['X-Total'])/registros) #Calculamos el número de páginas en base a todos los registros que hay en la API
            #Al dividir el número total de registros entre los registros solicitados obtenemos un aproximado, para asegurar que tomaremos todos, redondeamos al entero más alto
            for i in range(1,Iteraciones+1): #Ciclo for para obtener todos los registros de cada página
                url = f"https://api.pandascore.co/{juego}/{informacion}?page[size]={registros}&page[number]={i}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
                #URL el parámetro de página deseada
                response = requests.get(url) #Realizamos la petición correspondiente
                data = response.json() #Transformamos la respuesta 
                if i == 1: #En la primer iteración se crea un Dataframe
                    df_matches = pd.DataFrame.from_dict(data)
                else: #En el resto de iteraciones se van añadiendo los nuevos registros
                    df_matches = df_matches.append(pd.DataFrame.from_dict(data),ignore_index=True) 
            
            ########### Limpieza de datos
            campos_no_deseados = ['videogame', 'slug', 'tournament', 'league', 'serie', 'streams', 'modified_at', 'games',
                      'game_advantage', 'scheduled_at', 'original_scheduled_at', 'detailed_stats','live',
                      'match_type', 'videogame_version', 'live_embed_url', 'id'] 

            df_matches_temp = df_matches.drop(columns=campos_no_deseados) #Limpiamos campos que no nos son de utilidad
            df_matches_finished = df_matches_temp[ df_matches_temp['status'] == 'finished' ] #Dejamos solo los enfrentamientos que han terminado
            df_matches_finished_clean = df_matches_finished.dropna(subset=['begin_at', 'end_at']) #Eliminamos algunos valores nulos de las fechas
            df_matches_ordenado = df_matches_finished_clean[['name','opponents','results','forfeit','draw', #Reorganizamos los campos
                                                      'number_of_games','winner','winner_id','begin_at','end_at', 
                                                      'rescheduled','official_stream_url','league_id', 'serie_id', 'tournament_id']]
            df_matches_ordenado = df_matches_ordenado.reset_index(drop=True) #Reseteamos el índice
            nombres_nuevos = {'name' : 'match_name',
                             'begin_at' : 'begin_date',
                             'end_at' : 'end_date',
                             'official_stream_url' : 'official_stream_channel_url'}                       
            df_matches_ordenado.rename(columns=nombres_nuevos, inplace=True) #Renombramos columnas                                      
            return df_matches_ordenado # Finalmente se devuelve el Dataframe con todos los registros limpios y ordenados
        else: #Si la petición no se ejecutó correctamente mandamos un mensaje con el error devuelto
            return 'Error de petición: '+str(response.status_code)  
    else: #Si los argumentos no son de tipo str regresamos un mensaje
        return 'Por favor coloque un argumento válido en formato string.'

In [ ]:
import requests #Importamos los paquete necesarios
import pandas as pd 
import math as mt
 
def PandaScore_get_matches(juego): #Definimos nuestra función
    if isinstance(juego, str): #Nos aseguramos que los argumentos sean de tipo str
        informacion = 'matches'
        registros = 100 #Establecemos el número de registros por página
        token = 'V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns' #Token de acceso
        url = f"https://api.pandascore.co/{juego}/{informacion}?page[size]={registros}&token={token}" #URL con los parámetros deseados
        response = requests.get(url) #Realizamos la petición
        if response.status_code == 200: #Si todo salió bien continuamos
            Iteraciones = mt.ceil(float(response.headers['X-Total'])/registros) #Calculamos el número de páginas en base a todos los registros que hay en la API
            #Al dividir el número total de registros entre los registros solicitados obtenemos un aproximado, para asegurar que tomaremos todos, redondeamos al entero más alto
            for i in range(1,Iteraciones+1): #Ciclo for para obtener todos los registros de cada página
                url = f"https://api.pandascore.co/{juego}/{informacion}?page[size]={registros}&page[number]={i}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
                #URL el parámetro de página deseada
                response = requests.get(url) #Realizamos la petición correspondiente
                data = response.json() #Transformamos la respuesta 
                if i == 1: #En la primer iteración se crea un Dataframe
                    df_matches = pd.DataFrame.from_dict(data)
                else: #En el resto de iteraciones se van añadiendo los nuevos registros
                    df_matches = df_matches.append(pd.DataFrame.from_dict(data),ignore_index=True) 
            
            ########### Limpieza de datos
            campos_no_deseados = ['videogame', 'slug', 'tournament', 'league', 'serie', 'streams', 'modified_at', 'games',
                      'game_advantage', 'scheduled_at', 'original_scheduled_at', 'detailed_stats','live',
                      'match_type', 'videogame_version', 'live_embed_url', 'id'] 

            df_matches_temp = df_matches.drop(columns=campos_no_deseados) #Limpiamos campos que no nos son de utilidad
            df_matches_finished = df_matches_temp[ df_matches_temp['status'] == 'finished' ] #Dejamos solo los enfrentamientos que han terminado
            df_matches_finished_clean = df_matches_finished.dropna(subset=['begin_at', 'end_at']) #Eliminamos algunos valores nulos de las fechas
            df_matches_ordenado = df_matches_finished_clean[['name','opponents','results','forfeit','draw', #Reorganizamos los campos
                                                      'number_of_games','winner','winner_id','begin_at','end_at', 
                                                      'rescheduled','official_stream_url','league_id', 'serie_id', 'tournament_id']]
            df_matches_ordenado = df_matches_ordenado.reset_index(drop=True) #Reseteamos el índice
            nombres_nuevos = {'name' : 'match_name',
                             'begin_at' : 'begin_date',
                             'end_at' : 'end_date',
                             'official_stream_url' : 'official_stream_channel_url'}                       
            df_matches_ordenado.rename(columns=nombres_nuevos, inplace=True) #Renombramos columnas                                      
            return df_matches_ordenado # Finalmente se devuelve el Dataframe con todos los registros limpios y ordenados
        else: #Si la petición no se ejecutó correctamente mandamos un mensaje con el error devuelto
            return 'Error de petición: '+str(response.status_code)  
    else: #Si los argumentos no son de tipo str regresamos un mensaje
        return 'Por favor coloque un argumento válido en formato string.'

Le indicamos que deseamos los enfrentamientos del juego League of Legends:

In [ ]:
lol_matches = PandaScore_get_matches('lol')

Observamos el Dataframe que nos devuelve la función:

In [ ]:
lol_matches.head()

Es idéntico al Dataframe que habiamos elaborado en el postwork anterior, es notable que tiene un número de registros mayor, seguramente PandaScore recolecto más datos. Antes de comenzar a tratar los datos, nos gustaría modificar un poco la información que mostramos aprovechando 3 campos que dejamos en forma de lista:

*   _opponents_
*   _results_
*   _winner_

En oponentes tenemos los equipos que se enfrentaron, en results tenemos el puntaje de cada uno y en winner tenemos al equipo ganador. Proponemos entonces crear campos que muestren:

* winner_team
* winner_score
* loser_team
* loser_score

Para ello aprovecharemos las propiedades de listas y los aplicaremos a funciones:

In [ ]:
def winner_team(serie):
    if serie == None:
        return 'Undefined'
    else: return serie['name']

La función anterior devuelve el elemento ['name'] de la columna de listas _'winner'_ para aplicarla a toda la columna nos apoyamos del método map():

In [ ]:
lol_matches['winner_team'] = lol_matches['winner'].map(winner_team) #Recordemos que algunos enfrentamientos terminaron en empate

Para conocer a los equipos perdedores desglozamos el campo 'opponents' para que conocer a los equipos y comparamos con el equipo ganador previamente obtenido:

In [ ]:
def loser_team(df):
    losers = []
    for i in range(0,len(df)):
        ganador = df['winner_team'][i]
        equipo = df['opponents'][i][0]['opponent']['name']
        if ganador == 'Undefined':
           losers.append('Undefined')
        elif equipo == ganador:
           losers.append(df['opponents'][i][1]['opponent']['name'])  
        elif equipo != ganador: 
           losers.append(df['opponents'][i][0]['opponent']['name'] )
    return pd.Series(losers,index=df.index)

Ya que la función nos devuelve una serie de Pandas podemos añadir la nueva columna:

In [ ]:
lol_matches['loser_team'] = loser_team(lol_matches)

Antes de trabajar con las puntuaciones de los equipos, debemos realiar algunos cambios, previamente habiamos llegado al consenso de que en _winner id_ y _winner_ podrían existir valores _None_ debido a que había partidos con empate. Sin embardo para poder extraer los puntajes, es necesario cambiar los _Nones_ para así poder transformar el campo _winner id_ de un tipo de dato float a int y posteriormente realizar la extracción. Primero veamos si existe un _winner id_ que sea igual a cero para saber si podemos convertir los valores NaN en 0:

In [ ]:
lol_matches[ lol_matches['winner_id'] == 0]

No hay resultados, podemos transformar con seguridad los _None_ a 0, además corregiremos el campo '_official stream channel url_': 

In [ ]:
lol_matches['winner_id'] = lol_matches['winner_id'].fillna(0)
lol_matches['official_stream_channel_url'] = lol_matches['official_stream_channel_url'].fillna('Not Provided')

Veamos si se realizó correctamente el cambio:

In [ ]:
lol_matches['winner_id'].isna().sum()
lol_matches['official_stream_channel_url'].isna().sum()

¡Excelente! Ahora transformemos la columna a enteros:

In [ ]:
lol_matches['winner_id'] = lol_matches['winner_id'].astype(int)
lol_matches.info()

Ahora podemos crear nuestra función:

In [ ]:
def team_scores(df):
    losers_score = []
    winners_score = []
    for i in range(0,len(df)):
        ganador_id = df['winner_id'][i]
        equipo1_id = df['results'][i][0].get('team_id')
        equipo2_id = df['results'][i][1].get('team_id')
        if ganador_id == 0:
            winner_score = 0
            loser_score = 0
        elif equipo1_id == ganador_id:
            winner_score = df['results'][i][0].get('score')
            loser_score = df['results'][i][1].get('score')
        else:
            winner_score = df['results'][i][1].get('score')
            loser_score = df['results'][i][0].get('score')
        winners_score.append(winner_score)    
        losers_score.append(loser_score)
    return [pd.Series(winners_score,index=df.index), pd.Series(losers_score,index=df.index) ]

In [ ]:
[winner_score,loser_score] = team_scores(lol_matches)
lol_matches['winner_score'] = winner_score
lol_matches['loser_score'] = loser_score
lol_matches

Ahora que extrajimos la información más relevante de los campos con listas, podemos retirarlos:

In [ ]:
lol_matches2 = lol_matches
campos_no_deseados = ['results','winner','opponents','winner_id']
lol_matches2 = lol_matches2.drop(columns=campos_no_deseados)

In [ ]:
lol_matches2.info()

Nuestro siguiente objetivo es refinar los tipos de datos de los campos para las fechas:

In [ ]:
from datetime import *
lol_matches2['begin_date'] = lol_matches2['begin_date'].apply(datetime.strptime, args=('%Y-%m-%dT%H:%M:%SZ',))
lol_matches2['end_date'] = lol_matches2['end_date'].apply(datetime.strptime, args=('%Y-%m-%dT%H:%M:%SZ',))

Ya que la API guarda los datos en una forma algo peculiar nos apoyamos del paquete datetime y convertimos con ayuda de apply. Finalmente tendremos un Dataframe libre de valores nulos, con campos nombrados coherentementes (en inglés) y que ofrece la posibilidad de trabajar con otros datos de la API al conservar los identificadores:

In [ ]:
lol_matches2.info()

In [ ]:
lol_matches2.head()

Antes de guardar nuestro Dataframe, ordenemos por fecha de forma que lo primero que veamos sean los enfrentamientos más antiguos:

In [ ]:
lol_matches2 = lol_matches2.sort_values('begin_date', ascending=True)

Necesitamos resetear el índice:

In [ ]:
lol_matches2 = lol_matches2.reset_index(drop=True)
lol_matches2

Finalmente, guardamos el Dataframe:

In [ ]:
lol_matches2.to_csv('lol_matches2.csv')

# Postwork 08: Preparándose para las siguientes fases

Sea PandaScore_get_matches() la función que realiza peticiones a nuestra API sobre los enfrentamientos de un juego:

In [ ]:
import pandas as pd 
import math as mt
from datetime import *
import requests 

def PandaScore_get_matches(juego): #Definimos nuestra función
    if isinstance(juego, str): #Nos aseguramos que los argumentos sean de tipo str
        informacion = 'matches'
        registros = 100 #Establecemos el número de registros por página
        token = 'V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns' #Token de acceso
        url = f"https://api.pandascore.co/{juego}/{informacion}?page[size]={registros}&token={token}" #URL con los parámetros deseados
        response = requests.get(url) #Realizamos la petición
        if response.status_code == 200: #Si todo salió bien continuamos
            Iteraciones = mt.ceil(float(response.headers['X-Total'])/registros) #Calculamos el número de páginas en base a todos los registros que hay en la API
            #Al dividir el número total de registros entre los registros solicitados obtenemos un aproximado, para asegurar que tomaremos todos, redondeamos al entero más alto
            for i in range(1,Iteraciones+1): #Ciclo for para obtener todos los registros de cada página
                url = f"https://api.pandascore.co/{juego}/{informacion}?page[size]={registros}&page[number]={i}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
                #URL el parámetro de página deseada
                response = requests.get(url) #Realizamos la petición correspondiente
                data = response.json() #Transformamos la respuesta 
                if i == 1: #En la primer iteración se crea un Dataframe
                    df_matches = pd.DataFrame.from_dict(data)
                else: #En el resto de iteraciones se van añadiendo los nuevos registros
                    df_matches = df_matches.append(pd.DataFrame.from_dict(data),ignore_index=True)                    
            return df_matches # Finalmente se devuelve el Dataframe con todos los registros limpios y ordenados
        else: #Si la petición no se ejecutó correctamente mandamos un mensaje con el error devuelto
            return 'Error de petición: '+str(response.status_code)  
    else: #Si los argumentos no son de tipo str regresamos un mensaje
        return 'Por favor coloque un argumento válido en formato string.'

Además, sea df_matches_clean() la función que elimnará campos indeseados:

In [ ]:
def df_matches_clean(df_matches):
  ########### Limpieza de datos
   campos_no_deseados = ['videogame', 'slug', 'tournament', 'league', 'serie', 'streams', 'modified_at', 'games',
                      'game_advantage', 'scheduled_at', 'original_scheduled_at', 'detailed_stats','live',
                      'match_type', 'videogame_version', 'live_embed_url', 'id'] 

   df_matches_temp = df_matches.drop(columns=campos_no_deseados) #Limpiamos campos que no nos son de utilidad
   df_matches_finished = df_matches_temp[ df_matches_temp['status'] == 'finished' ] #Dejamos solo los enfrentamientos que han terminado
   df_matches_finished_clean = df_matches_finished.dropna(subset=['begin_at', 'end_at']) #Eliminamos algunos valores nulos de las fechas
   df_matches_ordenado = df_matches_finished_clean[['name','opponents','results','forfeit','draw', #Reorganizamos los campos
                                                      'number_of_games','winner','winner_id','begin_at','end_at', 
                                                      'rescheduled','official_stream_url','league_id', 'serie_id', 'tournament_id']]
   df_matches_ordenado = df_matches_ordenado.reset_index(drop=True) #Reseteamos el índice
   nombres_nuevos = {'name' : 'match_name',
                             'begin_at' : 'begin_date',
                             'end_at' : 'end_date',
                             'official_stream_url' : 'official_stream_channel_url'}                       
   df_matches_ordenado.rename(columns=nombres_nuevos, inplace=True) #Renombramos columnas 
   return df_matches_ordenado 

Junto a winner_team(),loser_team() y team_score() que añaden nuevos campos:

In [ ]:
def winner_team(serie):
    if serie == None:
        return 'Undefined'
    else: return serie['name']

def loser_team(df):
    losers = []
    for i in range(0,len(df)):
        ganador = df['winner_team'][i]
        equipo = df['opponents'][i][0]['opponent']['name']
        if ganador == 'Undefined':
           losers.append('Undefined')
        elif equipo == ganador:
           losers.append(df['opponents'][i][1]['opponent']['name'])  
        elif equipo != ganador: 
           losers.append(df['opponents'][i][0]['opponent']['name'] )
    return pd.Series(losers,index=df.index)

def team_scores(df):
    losers_score = []
    winners_score = []
    for i in range(0,len(df)):
        ganador_id = df['winner_id'][i]
        equipo1_id = df['results'][i][0].get('team_id')
        equipo2_id = df['results'][i][1].get('team_id')
        if ganador_id == 0:
            winner_score = 0
            loser_score = 0
        elif equipo1_id == ganador_id:
            winner_score = df['results'][i][0].get('score')
            loser_score = df['results'][i][1].get('score')
        else:
            winner_score = df['results'][i][1].get('score')
            loser_score = df['results'][i][0].get('score')
        winners_score.append(winner_score)    
        losers_score.append(loser_score)
    return [pd.Series(winners_score,index=df.index), pd.Series(losers_score,index=df.index) ]


Y finalmente sea matches_shiny() la función que embellecé nuestro dataframe y arregla los tipos de dato:

In [ ]:
def matches_shiny(juego_matches):
    juego_matches['winner_team'] = juego_matches['winner'].map(winner_team)
    juego_matches['loser_team'] = loser_team(juego_matches)  
    juego_matches['winner_id'] = juego_matches['winner_id'].fillna(0)
    juego_matches['official_stream_channel_url'] = juego_matches['official_stream_channel_url'].fillna('Not Provided')
    juego_matches['winner_id'] = juego_matches['winner_id'].astype(int)
    [winner_score,loser_score] = team_scores(juego_matches)
    juego_matches['winner_score'] = winner_score
    juego_matches['loser_score'] = loser_score
    campos_no_deseados = ['results','winner','opponents','winner_id']
    juego_matches = juego_matches.drop(columns=campos_no_deseados)
    juego_matches['begin_date'] = juego_matches['begin_date'].apply(datetime.strptime, args=('%Y-%m-%dT%H:%M:%SZ',))
    juego_matches['end_date'] = juego_matches['end_date'].apply(datetime.strptime, args=('%Y-%m-%dT%H:%M:%SZ',))
    ljuego_matches = juego_matches.sort_values('begin_date', ascending=True)
    juego_matches = juego_matches.reset_index(drop=True)
    return juego_matches

Llamamos a las funciones para obtener nuestro Dataframe de forma automática:

In [ ]:
lol_matches_raw = PandaScore_get_matches('lol')
lol_matches_sin_nones = df_matches_clean(lol_matches_raw)
lol_matches = matches_shiny(lol_matches_sin_nones)

lol matches se vuelve entonces uno de nuestros Dataframes destinado para el análisis de eSports:

In [ ]:
lol_matches

Gracias a nuestras funciones aportamos también otros Dataframe para los juegos de la API:

In [ ]:
dota2_matches_raw = PandaScore_get_matches('dota2')
dota2_matches_sin_nones = df_matches_clean(dota2_matches_raw)
dota2_matches = matches_shiny(dota2_matches_sin_nones)

In [ ]:
dota2_matches

In [ ]:
csgo_matches_raw = PandaScore_get_matches('csgo')
csgo_matches_sin_nones = df_matches_clean(csgo_matches_raw)
csgo_matches = matches_shiny(csgo_matches_sin_nones)

In [ ]:
csgo_matches

In [ ]:
ow_matches_raw = PandaScore_get_matches('ow')
ow_matches_sin_nones = df_matches_clean(ow_matches_raw)
ow_matches = matches_shiny(ow_matches_sin_nones)

In [ ]:
ow_matches

Podemos utilizar nuestro Dataframe para obtener datos interesantes que responden algunas preguntas que nos hicimos al comienzo. En un principio nos preguntabamos si los enfrentamientos de eSports son cortos o largos, aprovechemos el paquete _datetime_ y descubrámoslo:

In [ ]:
def tiempo_partida(df):
    minutos_partidas=[] #iniciamos un arreglo vacío
    for n in range (0,len(df)):   
        inicio_enfrentamiento= df.loc[n,'begin_date'] #obtenemos el momento en que inici´p el enfrentamiento
        fin_enfrentamiento= df.loc[n,'end_date'] #obtenemos el momento en que terminó el enfrentamiento
        total_partidas = df.loc[n,'number_of_games'] #obtenemos el número de partidas jugadas en el enfrentamiento
        duración_partidas = (fin_enfrentamiento-inicio_enfrentamiento)/total_partidas #Calculamos el tiempo por partida
        minutos = round(duración_partidas.total_seconds()/60,3) # Convertimos a segundos y de segundos a minutos
        minutos_partidas.append(minutos) #Añadimos al arreglo
  
    df['minutos_partidas'] = pd.Series(minutos_partidas) #Finalmente creamos una nueva columna con los tiempos por partida
    no_outlier = df['minutos_partidas'] > 0 # La API devovió algunos datos de fechas con errores donde end_date era antes de begin_date
    df= df[no_outlier]
    no_outlier = df['minutos_partidas'] < 100 # Sabemos por experiencia que las partidas no duran tanto, así que eliminamos outliers
    df_boxplot = df[no_outlier]   
          
    return df_boxplot #devolvemos un dataframe filtrado con la duración de partidas en minutos

La función anterior esta pensada para obtener la duración en minutos de las partidas, se nos devuelve el dataframe que usamos como argumento con nuevo campo y filtrado para partidas con tiempo mayor a 0 (tiempo positivo) y menor a 100 minutos (como jugones curtidos sabemos que 100 minutos es una partida exagerada por si sola). Veamos como se comportan los diferentes juegos con un gráfico llamativo que podemos generar gracias al paquete seaborn:

In [ ]:
#Creamos los dataframe con las funciones
boxplot_lol = tiempo_partida(lol_matches)
boxplot_dota2 = tiempo_partida(dota2_matches)
boxplot_csgo = tiempo_partida(csgo_matches)
boxplot_ow = tiempo_partida(ow_matches)
#llamamos el paquete seaborn
import seaborn as sns
#Diseñamos la figura
plt.figure(figsize=[20, 10])
sns.set(style="whitegrid")
sns.boxplot(data=[boxplot_ow['minutos_partidas'],boxplot_lol['minutos_partidas'],boxplot_dota2['minutos_partidas'],boxplot_csgo['minutos_partidas']]).set_xticklabels(['OW','LoL','Dota2','CS:GO'])
plt.title("Duración de partidas en torneos de Esports")
plt.ylabel("Minutos")
plt.xlabel("Juegos")
plt.show()

Analizando los resultados de manera muy rápida, encontramos que las medianas rondan en torno a los 25 minutos, 40 minutos, 45 minutos y 50 minutos respectivamente. Si esto llegan a tardar los profesionales, qué nos espera a nosotros los mortales? Muy bien, ahora veamos las proporciones de enfrentamientos que han sido transmitidos en línea para que los fanáticos los vean:

In [ ]:
# Obtendremos las proporciones manualmente utilizando el tamaño de los Dataframe y filtrando los streams
Total_de_partidas_lol = len(lol_matches['official_stream_channel_url']) #Total de enfrentamientos 
Partidas_sin_stream_lol = len(lol_matches[ lol_matches['official_stream_channel_url'] == 'Not Provided']) #Total de enfrentamientos sin un url de stream
Partidas_con_stream_lol = Total_de_partidas_lol-Partidas_sin_stream_lol #Total de enfrentamientos que cuentan con url para stream
data = [Partidas_con_stream_lol,Partidas_sin_stream_lol] #Creamos un arreglo 
streams_lol = pd.Series(data, index = ['Transmitidas en línea','No transmitidas']) #Creamos una serie para el gráfico de tipo pie 

Total_de_partidas_dota2 = len(dota2_matches['official_stream_channel_url'])
Partidas_sin_stream_dota2 = len(dota2_matches[ dota2_matches['official_stream_channel_url'] == 'Not Provided'])
Partidas_con_stream_dota2= Total_de_partidas_dota2-Partidas_sin_stream_dota2
data = [Partidas_con_stream_dota2,Partidas_sin_stream_dota2]
streams_dota2 = pd.Series(data, index = ['Transmitidas en línea','No transmitidas'])

Total_de_partidas_csgo = len(csgo_matches['official_stream_channel_url'])
Partidas_sin_stream_csgo = len(csgo_matches[ csgo_matches['official_stream_channel_url'] == 'Not Provided'])
Partidas_con_stream_csgo= Total_de_partidas_csgo-Partidas_sin_stream_csgo
data = [Partidas_con_stream_csgo,Partidas_sin_stream_csgo]
streams_csgo= pd.Series(data, index = ['Transmitidas en línea','No transmitidas'])

Total_de_partidas_ow = len(ow_matches['official_stream_channel_url'])
Partidas_sin_stream_ow= len(ow_matches[ ow_matches['official_stream_channel_url'] == 'Not Provided'])
Partidas_con_stream_ow = Total_de_partidas_ow-Partidas_sin_stream_ow
data = [Partidas_con_stream_ow,Partidas_sin_stream_ow]
streams_ow = pd.Series(data)

La mejor forma de ver las proporciones es con un gráfico de pastel, nos apoyaremos en el paquete matplotlib para esta tarea:

In [ ]:
# llamamos al paquete matplotlib
import matplotlib.pyplot as plt 
# declaramos las etiquetas
labels = ['Transmitidas en línea','No transmitidas']
# paletas de colores https://www.w3schools.com/colors/colors_names.asp
fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2) # creamos un subplot de 2x2
fig.suptitle('Proporción de partidas de eSports trasmitidas por internet', fontsize = 30) # añadimos un título
fig.set_figheight(15) #dimensiones de la figura
fig.set_figwidth(15)
# gráfico de pastel para las proprociones de cada juego y su título
ax1.pie(x=streams_lol, autopct="%.1f%%", labels=labels, pctdistance=0.5, colors= ('DarkOliveGreen','DarkGreen')) 
ax1.set_title('League of Legends')
ax2.pie(x=streams_dota2, autopct="%.1f%%", labels=labels, pctdistance=0.5, colors= ('DarkMagenta','DarkOrchid'))
ax2.set_title('Dota 2')
ax3.pie(x=streams_csgo, autopct="%.1f%%", labels=labels, pctdistance=0.5, colors= ('Coral','Bisque'))
ax3.set_title('Counter Strike: Global Offensive')
ax4.pie(x=streams_ow, autopct="%.1f%%", labels=labels, pctdistance=0.5, colors= ('b','c'))
ax4.set_title('Overwatch')

En general tenemos un gráfico fácil de leer, los enfrentamientos para cada juego con transmitidos en más del 50% de los casos, siendo LoL el de menor proporción y CS:GO el de mayor proporción. 

Queda comprobado entonces que nuestros Dataframes forman un Dataset con gran potencial de explotación y con el que podemos obtener mucha información relacionada a eSports, pero aún nos falta responder una última pregunta para quedar satisfechos.